File: packages/api/op_agent/repo/memory.py
"""記憶體版資料層。

程序重啟資料就消失，只用於本地開發 / demo 前的流程驗證。
Flask debug reloader 會重新 import module，所以資料存在 module 級變數即可。
"""

from __future__ import annotations

import copy

from ..domain import (
    Booking,
    ChatSession,
    MatchResult,
    ServiceRequest,
    UserPreferences,
    UserProfile,
    Vendor,
)
from ..seed import SEED_USERS, SEED_VENDORS
from .base import merge_prefs, vendor_covers


class MemoryRepo:
    def __init__(self) -> None:
        self._users: dict[str, UserProfile] = {
            u["inbrAccountId"]: copy.deepcopy(u) for u in SEED_USERS
        }
        self._vendors: dict[str, Vendor] = {
            v["vendorId"]: copy.deepcopy(v) for v in SEED_VENDORS
        }
        self._requests: dict[str, ServiceRequest] = {}
        self._matches: dict[str, MatchResult] = {}
        self._bookings: dict[str, Booking] = {}
        self._sessions: dict[str, ChatSession] = {}

    # ---- 會員 ----
    def get_user(self, inbr_account_id: str) -> UserProfile | None:
        return self._users.get(inbr_account_id)

    def put_user(self, user: UserProfile) -> None:
        self._users[user["inbrAccountId"]] = user

    def merge_preferences(self, inbr_account_id: str, patch: UserPreferences) -> UserPreferences:
        user = self._users.get(inbr_account_id)
        if user is None:
            raise KeyError(f"user not found: {inbr_account_id}")
        user["preferences"] = merge_prefs(user.get("preferences", {}), patch)
        return user["preferences"]

    # ---- 廠商 ----
    def list_vendors(
        self,
        *,
        category: str | None = None,
        county_code: str | None = None,
        district_code: str | None = None,
    ) -> list[Vendor]:
        items = list(self._vendors.values())
        if category:
            items = [v for v in items if category in v.get("categories", [])]
        if county_code:
            items = [v for v in items if vendor_covers(v, county_code, district_code)]
        return items

    def get_vendor(self, vendor_id: str) -> Vendor | None:
        return self._vendors.get(vendor_id)

    def put_vendor(self, vendor: Vendor) -> None:
        self._vendors[vendor["vendorId"]] = vendor

    # ---- 服務單 ----
    def get_request(self, request_id: str) -> ServiceRequest | None:
        return self._requests.get(request_id)

    def put_request(self, req: ServiceRequest) -> None:
        self._requests[req["requestId"]] = req

    def list_requests_by_user(self, inbr_account_id: str) -> list[ServiceRequest]:
        return [r for r in self._requests.values() if r["inbrAccountId"] == inbr_account_id]

    # ---- 媒合結果 ----
    def put_match(self, match: MatchResult) -> None:
        self._matches[match["requestId"]] = match

    def get_match(self, request_id: str) -> MatchResult | None:
        return self._matches.get(request_id)

    # ---- 預約單 ----
    def put_booking(self, booking: Booking) -> None:
        self._bookings[booking["orderNo"]] = booking

    def get_booking(self, order_no: str) -> Booking | None:
        return self._bookings.get(order_no)

    def list_bookings_by_user(self, inbr_account_id: str) -> list[Booking]:
        return [b for b in self._bookings.values() if b["inbrAccountId"] == inbr_account_id]

    # ---- 對話 ----
    def get_session(self, session_id: str) -> ChatSession | None:
        return self._sessions.get(session_id)

    def put_session(self, session: ChatSession) -> None:
        self._sessions[session["sessionId"]] = session

File: packages/backend/src/repo/memory.ts
import type {
  Booking,
  ChatSession,
  MatchResult,
  ServiceCategory,
  ServiceRequest,
  UserPreferences,
  UserProfile,
  Vendor,
} from '../domain/types';
import { SEED_USERS } from '../data/users';
import { SEED_VENDORS } from '../data/vendors';
import type { Repo } from './types';
import { vendorCovers } from './coverage';

/**
 * 記憶體版資料層。程序重啟資料就消失，只用於本地開發 / demo 前的流程驗證。
 * 用 globalThis 存，這樣 tsx watch 熱重載時對話不會被清掉。
 */
interface Store {
  users: Map<string, UserProfile>;
  vendors: Map<string, Vendor>;
  requests: Map<string, ServiceRequest>;
  matches: Map<string, MatchResult>;
  bookings: Map<string, Booking>;
  sessions: Map<string, ChatSession>;
}

const g = globalThis as unknown as { __opStore?: Store };

function store(): Store {
  if (!g.__opStore) {
    g.__opStore = {
      users: new Map(SEED_USERS.map((u) => [u.inbrAccountId, structuredClone(u)])),
      vendors: new Map(SEED_VENDORS.map((v) => [v.vendorId, structuredClone(v)])),
      requests: new Map(),
      matches: new Map(),
      bookings: new Map(),
      sessions: new Map(),
    };
  }
  return g.__opStore;
}

export class MemoryRepo implements Repo {
  async getUser(id: string) {
    return store().users.get(id);
  }

  async putUser(user: UserProfile) {
    store().users.set(user.inbrAccountId, user);
  }

  async mergePreferences(id: string, patch: UserPreferences): Promise<UserPreferences> {
    const user = store().users.get(id);
    if (!user) throw new Error(`user not found: ${id}`);
    user.preferences = mergePrefs(user.preferences, patch);
    return user.preferences;
  }

  async listVendors(filter?: {
    category?: ServiceCategory;
    countyCode?: string;
    districtCode?: string;
  }) {
    let list = [...store().vendors.values()];
    if (filter?.category) list = list.filter((v) => v.categories.includes(filter.category!));
    if (filter?.countyCode) {
      list = list.filter((v) => vendorCovers(v, filter.countyCode!, filter.districtCode));
    }
    return list;
  }

  async getVendor(vendorId: string) {
    return store().vendors.get(vendorId);
  }

  async putVendor(vendor: Vendor) {
    store().vendors.set(vendor.vendorId, vendor);
  }

  async getRequest(requestId: string) {
    return store().requests.get(requestId);
  }

  async putRequest(req: ServiceRequest) {
    store().requests.set(req.requestId, req);
  }

  async listRequestsByUser(id: string) {
    return [...store().requests.values()].filter((r) => r.inbrAccountId === id);
  }

  async putMatch(match: MatchResult) {
    store().matches.set(match.requestId, match);
  }

  async getMatch(requestId: string) {
    return store().matches.get(requestId);
  }

  async putBooking(booking: Booking) {
    store().bookings.set(booking.orderNo, booking);
  }

  async getBooking(orderNo: string) {
    return store().bookings.get(orderNo);
  }

  async listBookingsByUser(id: string) {
    return [...store().bookings.values()].filter((b) => b.inbrAccountId === id);
  }

  async getSession(sessionId: string) {
    return store().sessions.get(sessionId);
  }

  async putSession(session: ChatSession) {
    store().sessions.set(session.sessionId, session);
  }
}

/** 偏好合併規則：陣列做去重聯集、數值直接覆寫、notes 累加 */
export function mergePrefs(base: UserPreferences, patch: UserPreferences): UserPreferences {
  const uniq = <T>(a: T[] = [], b: T[] = []) => [...new Set([...a, ...b])];
  return {
    priceSensitivity: patch.priceSensitivity ?? base.priceSensitivity,
    preferredContactTime: patch.preferredContactTime ?? base.preferredContactTime,
    preferredVendorTags: uniq(base.preferredVendorTags, patch.preferredVendorTags),
    blockedVendorIds: uniq(base.blockedVendorIds, patch.blockedVendorIds),
    interestedCategories: uniq(base.interestedCategories, patch.interestedCategories),
    notes: uniq(base.notes, patch.notes).slice(-20),
  };
}
